File: solutions/fortran/collatz-conjecture/1/collatz_conjecture.f90
module collatz_conjecture
  implicit none
contains

  integer function steps(i)
    integer :: i
    integer :: n

    steps = -1
    if (i <= 0) stop

    steps = 0
    n = i
    do while ( n > 1 )
       if ( mod(n, 2) == 0 ) then
          n = n / 2
       else
          n = n * 3 + 1
       end if
       steps = steps + 1
    end do

  end function

end module

File: solutions/c/collatz-conjecture/1/collatz_conjecture.c
#include "collatz_conjecture.h"

int steps(int start) {
    if ( start < 1 ) { return ERROR_VALUE; }

    int count  = 0;
    int number = start;

    while ( number > 1 ) {
        count++;
        number = ((number & 1) == 0) ? number / 2 : 3 * number + 1;
    }

    return count;
}
